### PR TITLE
storaged: support creating btrfs snapshots

### PR DESCRIPTION
**Pull request by jelly**
Btrfs supports creating a snapshot of a subvolume, either readonly or write-able. Unlike creating subvolumes, it doesn't make sense to put a snapshot of a subvolume in the subvolume itself users likely have a special `snapshots` subvolume where they collect their snapshots.

Open questions:
* Unlike subvolumes we don't want to create a snapshot in the parent directory, I have given the UI an option to specify a full path, this is good and bad. Good because it is flexible, bad because it would allow you to specify a path to a NFS mount, xfs partition or something else weird. Then again it isn't too bad, note that snapshots can be created under a subvolume or normal directory in btrfs. (Users usually have a snapshot subvolume directory where they put them either `/snapshots or `/.snapshots`, this is not a well-known directory)
* Do we offer mount options? You can obviously mount them like subvolumes as they are one, so we can offer that flexibility

Follow up / issues:
* We don't identify snapshots, we just show it as subvolume
* We don't and likely can't identify read only snapshots, this requires major udisks work.

---

# Support creating btrfs snapshots

Screenshot from 2024-06-07 11-36-05

Snapshot creation dialog:

image

### CODE REVIEW

**Comment by garrett** [comment]
When you're creating snapshots, you do not want to mount them, or set up auto mounting? I don't understand why mounting is half the dialog.

Snapshots of subvolumes are intended as "cheap" "backups". I don't even remember mounting or unmounting them when using them before in the past. They also cannot cross into other subvolumes. I don't think they're even treated like subvolumes in a traditional manner.

Isn't btrfs snapshotting basically the following:

- the volume you're wanting to snapshot
- the location _within that same subvolume_ that you're snapshotting
- if it's a read-only or read-write snapshot
- optionally, quota group(s)

Nothing related to mounting or even unmounting.

(Whenever I've made a snapshot, I'm able to go into that snapshot immediately without having to do anything special.)

If someone wants to mount a snapshot instead of going into it, they can do that with the standard mount dialog, right?

**Comment by garrett** [comment]
So I was thinking we can make snapshotting more like a snapshotting task and less like "random generic subvolume". This would include some nice stuff where it makes snapshotting super easy without "crap work" that someone would have to do each time, and have the computer remember and figure out that stuff for them, with good and opinionated defaults.

Here's what that could look like:

create snapshot dialog

The text in the side says:

- Save the snapshot location for each subvolume in local storage, so when someone comes back to create a new snapshot, it'll be the same as before
- Remember the last selection for snapshot name too
- Selecting custom shows a text entry

**Comment by Conan-Kudo** [comment]
If you're trying to snapshot the parent subvolumes (e.g. `root` and `home` in Fedora's default setup), you will want to consider putting those snapshots in the parent volume that's not normally mounted (that is, the actual base btrfs volume).

Put more concretely, in Fedora's setup, we have the following setup:

```
Unmounted base volume
|
|- root
|- home
```

You could do something like this:

```
Unmounted base volume
|
|- root
|- home
|- .snapshots
|-|
|-|- root-20240615
```

This would allow you to have the snapshots not visible most of time, and if you need it for recovery or boot-to-snapshot purposes, it's easily available.

**Comment by jelly** [comment]
> If you're trying to snapshot the parent subvolumes (e.g. `root` and `home` in Fedora's default setup), you will want to consider putting those snapshots in the parent volume that's not normally mounted (that is, the actual base btrfs volume).
>
> Put more concretely, in Fedora's setup, we have the following setup:
> 
> ```
> Unmounted base volume
> |
> |- root
> |- home
> ```
> 
> You could do something like this:
> 
> ```
> Unmounted base volume
> |
> |- root
> |- home
> |- .snapshots
> |-|
> |-|- root-20240615
> ```
> 
> This would allow you to have the snapshots not visible most of time, and if you need it for recovery or boot-to-snapshot purposes, it's easily available.

The Cockpit storage page (currently) shows all subvolumes not based on what is mounted (that is an interesting idea though). There is a draft PR to hide snapshots from the overview page and list them under the subvolume you made a snapshot of. https://github.com/cockpit-project/cockpit/pull/20234

**Comment by mvollmer** [comment]
> I have given the UI an option to specify a full path, this is good and bad. 

Is this a path in the filesystem? I think it should be the name of the parent subvolume.

**Comment by jelly** [comment]
* Snapshot location is remembered per subvolume
* Directory Autocomplete (filter files), with an option to create a new subvolume to put snapshots. (a helper text should show that)
* Validate if specify a location which is not on the same btrfs volume

File: pkg/storaged/btrfs/subvolume.jsx [hunk 1]
```diff
@@ -180,6 +181,149 @@ function subvolume_create(volume, subvol, parent_dir) {
...
 
+async function snapshot_create(volume, subvol, subvolume_path) {
+    const localstorage_key = "storage:snapshot-locations";
+    console.log(volume, subvol, subvolume_path);
+    const action_variants = [
+        { tag: null, Title: _("Create snapshot") },
+    ];
+
+    const get_local_storage_snapshots_locs = () => {
+        const localstorage_snapshot_data = localStorage.getItem(localstorage_key);
```
Comment (mvollmer) [inline pkg/storaged/btrfs/subvolume.jsx:192]: What about using the location of the most recently made snapshot? By looking at the list of snapshots of this subvolume. That makes this independent from browser storage.

File: pkg/storaged/btrfs/subvolume.jsx [hunk 2]
```diff
@@ -180,6 +181,149 @@ function subvolume_create(volume, subvol, parent_dir) {
...
+                          value: subvol.pathname,
+                          disabled: true,
+                      }),
+            TextInput("snapshots_location", _("Snapshots location"),
+                      {
+                          value: get_localstorage_snapshot_location(subvol),
+                          placeholder: cockpit.format(_("Example, $0"), "/.snapshots"),
+                          explanation: (<>
+                              <p>{_("Snapshots must reside within the same btrfs volume.")}</p>
+                              <p>{_("When the snapshot location does not exist, it will be created as btrfs subvolume automatically.")}</p>
```
Comment (mvollmer) [inline pkg/storaged/btrfs/subvolume.jsx:276]: Hmm, why not as a directory? Also, non existing parents should be created as well, no?

File: pkg/storaged/btrfs/subvolume.jsx [hunk 3]
```diff
@@ -180,6 +181,149 @@ function subvolume_create(volume, subvol, parent_dir) {
...
+
+    dialog_open({
+        Title: _("Create snapshot"),
+        Fields: [
+            TextInput("subvolume", _("Subvolume"),
+                      {
+                          value: subvol.pathname,
+                          disabled: true,
+                      }),
+            TextInput("snapshots_location", _("Snapshots location"),
```
Comment (mvollmer) [inline pkg/storaged/btrfs/subvolume.jsx:270]: I think this should be a autocomplete thing, offering the mount points of all mounted subvolumes.

File: pkg/storaged/btrfs/verify-btrfs-snapshot-location.py
Comment (mvollmer) [inline pkg/storaged/btrfs/verify-btrfs-snapshot-location.py:1]: This seems unused.

**Review by mvollmer — CHANGES_REQUESTED** [review]

**Comment by Conan-Kudo** [comment]
Is there any chance that this would be picked up again? [Fedora Cloud now has `/boot` on Btrfs](https://fedoraproject.org/wiki/Changes/BtrfsBootForCloud) and @davdunc is interested in exposing/promoting Cockpit as the preferred in-OS interface model for this sort of thing.

**Comment by Conan-Kudo** [comment]
Also FYI @bmr-cymru

**Comment by bmr-cymru** [comment]
Thanks @Conan-Kudo! I'm on PTO right now, but I'll be back on Tuesday and I'll take a more thorough look then.

Just to comment quickly & in general: I guess I would quite like to see this functionality subsumed into Cockpit support for `snapm` (since we handle different snapshot backends, multiple volumes, naming, and a bunch of other stuff that frees the user from having to think about it).

We have a Python API and we are actively looking at a possible DBus one (or other RPC: potentially `varlink` or Cap'n Proto - we are evaluating options and fit, but DBus has a lot going for it in terms of ubiquity already). It may make sense for Cockpit support to wait for that since it would solve a bunch of seemingly relevant problems. If this work wanted to go ahead regardless without waiting on us I wouldn't have any objection to that.

I'll try to answer more usefully/completely next week.

**Comment by jelly** [comment]
> We have a Python API and we are actively looking at a possible DBus one (or other RPC: potentially varlink or Cap'n Proto - we are evaluating options and fit, but DBus has a lot going for it in terms of ubiquity already). It may make sense for Cockpit support to wait for that since it would solve a bunch of seemingly relevant problems. If this work wanted to go ahead regardless without waiting on us I wouldn't have any objection to that.

In short, DBus is totally fine and is what we already use for most things (udisks, systemd, etc.). Varlink should be fine as a well, we haven't used it anywhere yet as it is quite "new" but it is simple enough to not be a problem.
Cap’n Proto is something I have never heard of, not sure how we would interface with it from Cockpit and if it would require a new dependency.
Alternatively if the CLI has JSON output, Cockpit can also simply invoke for example `snapm snapset list --output=json` and parse that.

**Comment by bmr-cymru** [comment]
> In short, DBus is totally fine and is what we already use for most things (udisks, systemd, etc.). Varlink should be fine as a well, we haven't used it anywhere yet as it is quite "new" but it is simple enough to not be a problem. Cap’n Proto is something I have never heard of, not sure how we would interface with it from Cockpit and if it would require a new dependency. Alternatively if the CLI has JSON output, Cockpit can also simply invoke for example `snapm snapset list --output=json` and parse that.

I should probably elaborate a bit here on our actual thinking as it's definitely relevant for other components wanting to interface with us in future.

I favour simplicity and I would *like* to have one RPC to rule them all if possible. DBus is definitely the "natural" choice here - but - I have some concerns about how this would work for the application coordination feature we are discussing. This is a proposal for a  kind of "rendezvous" mechanism that would flow roughly like this (sorry, I'm not good at the modern `mermaid` style sequence diagrams; bullets will do for now):

* Apps register interest in snapshots, and the volumes they care about
* `snapm` keeps track of registered apps and their volumes of interest
* A user requests a new snapshot set for volumes X, Y, Z
  * `snapm` notifies interested parties for X, Y, Z, "we are about to make a snapshot"
  * Apps ack back to `snapm` and vote on whether to go ahead
  * If vote succeeds, `snapm` sends "about to snapshot - quiesce your on-disk data" message
  * Apps do their thing, second round of voting to confirm application consistent image exists
  * If vote succeeds `snapm` takes the snapshots
  * `snapm` tells everyone "all done, as you were"
  * Apps un-freeze and continue

This is a simplification  obviously and ignores important details like error handling, unregistration, etc. but it is conceptually similar to what [Microsoft VSS](https://learn.microsoft.com/en-us/windows/win32/vss/volume-shadow-copy-service-overview) has been doing for many years.

My worry is that doing this in DBus could be clumsy and slow - every client app would need to publish a service (I think - I am not an expert here...), and that could get messy. [Cap'n Proto](https://capnproto.org/)  has some nice built in features that seem to suit this domain better, and reportedly has very good performance.

All that said - this feature is still a ways out, and is orthogonal to our offering the current `Manager` API over something like DBus or Varlink. I think I need to accept that if/when we do go down that route, it may be necessary to use something other than those two, and I think that's fine - right tools for the job.

I'm on PTO right now  (back tomorrow), so this is a bit hasty and off-the-cuff. Let me know if you have any questions or if I can clarify anything.

Tl;dr I think we will probably end up offering a simple DBus or Varlink service at some point, although I can't say exactly when that might happen right now.

**Comment by bmr-cymru** [comment]
On the JSON point - yes - we do support JSON output. There are two formats: one based on the report system (`list` commands, where you can choose fields and sort order), and one based on the `show` output, which uses fixed fields. We plan to publish formal schemas for both of these in future but it's not done yet. We do consider them stable though.

```text
root@f42-snapm-vm1:~/src/git/snapm# snapm snapset list --json -osnapset_all
{
    "Snapsets": [
        {
            "snapset_name": "before-upgrade",
            "snapset_uuid": "b6a0adc0-138b-5656-bdf9-bedb2612de4e",
            "snapset_basename": "before-upgrade",
            "snapset_index": "-",
            "snapset_timestamp": 1765542336,
            "snapset_time": "2025-12-12 12:25:36",
            "snapset_nr_snapshots": 2,
            "snapset_sources": [
                "/",
                "/var"
            ],
            "snapset_mountpoints": [
                "/",
                "/var"
            ],
            "snapset_devices": [
                ""
            ],
            "snapset_status": "Active",
            "snapset_autoactivate": true,
            "snapset_mounted": false,
            "snapset_origin_mounted": true,
            "snapset_mount_root": "",
            "snapset_bootable": false,
            "snapset_bootentry": "",
            "snapset_revertentry": "",
            "snapset_categories": [
                "yearly",
                "monthly",
                "daily",
                "hourly"
            ]
        },
        {
            "snapset_name": "just-now",
            "snapset_uuid": "00f5b0ba-c326-5e6d-92eb-82ce8601b369",
            "snapset_basename": "just-now",
            "snapset_index": "-",
            "snapset_timestamp": 1768581289,
            "snapset_time": "2026-01-16 16:34:49",
            "snapset_nr_snapshots": 3,
            "snapset_sources": [
                "/",
                "/home",
                "/var"
            ],
            "snapset_mountpoints": [
                "/",
                "/home",
                "/var"
            ],
            "snapset_devices": [
                ""
            ],
            "snapset_status": "Active",
            "snapset_autoactivate": true,
            "snapset_mounted": false,
            "snapset_origin_mounted": true,
            "snapset_mount_root": "",
            "snapset_bootable": true,
            "snapset_bootentry": "3c7245835d50ac82947922902a6cb466814942c7",
            "snapset_revertentry": "f6569cdeaedf38ca543ebfe51ab860334ff05a93",
            "snapset_categories": [
                "yearly",
                "quarterly",
                "monthly",
                "daily",
                "hourly"
            ]
        }
    ]
}
```

```text
root@f42-snapm-vm1:~/src/git/snapm# snapm snapset show --json just-now
[
    {
        "SnapsetName": "just-now",
        "Sources": [
            "/",
            "/var",
            "/home"
        ],
        "MountPoints": [
            "/",
            "/var",
            "/home"
        ],
        "Devices": [],
        "NrSnapshots": 3,
        "Timestamp": 1768581289,
        "Time": "2026-01-16 16:34:49",
        "UUID": "00f5b0ba-c326-5e6d-92eb-82ce8601b369",
        "Status": "Active",
        "Categories": [
            "yearly",
            "quarterly",
            "monthly",
            "daily",
            "hourly"
        ],
        "Autoactivate": true,
        "OriginMounted": true,
        "Mounted": false,
        "Bootable": true,
        "BootEntries": {
            "SnapshotEntry": "3c72458",
            "RevertEntry": "f6569cd"
        }
    }
]
```

HTH!
